Paper: dagster-slurm: Dagster Assets on Slurm-Managed Scientific Workflows (#1259)
scipy-conference/scipy_proceedings: papers/dagster_slurm/main.md

Reviewer: The paper is really good and strong. Bringing Slurm managed HPC under Dagster which runs on local, CI and production cluster is perfect. The evidences provided with real production deployment is a great contribution and add on to the paper

Reviewer: The overall pdf when I download and read is around 12 pages without references. The guidance mentioned 8 pages, some repetitive sections can be concise

Reviewer: This paper is highly technical, and most of the acronyms are commonly known, such as SSH, GPU etc.. but still per the guideline some acronyms can be expanded in the first use to easily follow the paper... such as QoS, VAT identifiers etc. A non-European reader might not understand DACH as well, so any acronyms that are not very common can be expanded

  We present dagster-slurm, an open-source package that extends Dagster's control plane to Slurm-managed clusters.
  dagster-slurm lets the same asset source code run on a laptop, in continuous integration, against a containerized Slurm cluster, and on a production HPC cluster reached over SSH, while target-specific queues, credentials, filesystem paths, and site defaults remain in deployment configuration.
  A `ComputeResource` abstraction packages environments, submits jobs through Slurm, streams logs through Dagster Pipes, and records scheduler metadata such as job state, CPU efficiency, memory use, and node-hours.

  We evaluate the package through a production pipeline that discovers and characterizes family-owned firms across European countries from Common Crawl and a commercial company registry.
  The pipeline spans four compute tiers, from a DuckDB index scan over roughly 170 million URLs per crawl to fine-tuned language-model extraction on A100 GPUs.
  On the TU Wien DataLAB cluster, the pipeline has run more than 9,600 Slurm jobs while preserving the same Dagster asset graph used on a developer laptop and in CI.

>   We also report portability checks using site configurations for the VSC-5 and Leonardo supercomputers.

Reviewer: This line here is in here part of the production evaluation, however table 2 says no production workloads ran, and the operational experience section says the development integration was run on VSC5 and Leonardo but the results are no longer available. The claims sound inconsistent, please align this.

The contribution here is different: Slurm becomes one execution target for a Dagster asset graph that also touches object stores, databases, local development, and CI.
Teams that already use task workflow managers for compute-heavy stages can still wrap those invocations inside Dagster assets to recover lineage over the full dataflow.

We evaluate the integration with a production pipeline that discovers family-owned firms across European countries from web-scale text and matches them to a commercial company registry.
Throughout the paper, **tier** means a class of hardware matched to a pipeline stage, such as CPU filtering, CPU-plus-network fetching, GPU training, or GPU inference.
An **execution target** is a place the same asset source can run, such as a laptop, CI, containerized Slurm, or a production cluster.
The case-study pipeline exercises four compute tiers and scans roughly 170 million URLs per crawl across dozens of crawls.
We use it to make the system design concrete; the domain results are descriptive artifacts of the case study rather than validated empirical claims.

## System design

> The integration is organized in three layers, shown below, and mapped onto execution targets in @fig:arch.

Reviewer 2: ```suggestion
The integration is organized in three layers (@fig:layers) and mapped onto execution targets in @fig:arch.
```

@fig:graph combines the asset graph, compute tiers, scheduler metadata, execution targets, and measured Slurm totals for the case-study deployment.

> :::{figure} software_layers.png
> The three software layers.

Reviewer 2: ```suggestion
:::{figure} software_layers.png
:label: fig:layers
The three software layers.
```

The recommended pattern is a deployment-mode-aware path that resolves to a local directory during development and to a shared parallel filesystem or object store in production, so that asset code stays unchanged across targets.
This matters because data-intensive HPC workloads such as ours operate on data large enough that automatic serialization across network boundaries would be impractical.

### Environment packaging and reproducibility boundaries

Remote execution requires the cluster to have the same software the developer used.
dagster-slurm packages environments with Pixi and pixi-pack [@pixi]: a Pixi lockfile pins every dependency, and pixi-pack produces a relocatable bundle that is uploaded and unpacked on the target, or, on sites where a shared environment is pre-deployed, reused so that only the per-run payload script is transferred.
This addresses repeatability of the execution environment, but full scientific reproducibility of a production run additionally depends on inputs and state the package does not control: the specific Common Crawl snapshots, the licensed registry data, the model checkpoints, site-specific modules and CUDA drivers, and credentials.

> We therefore frame the reproducibility contribution as repeatable, version-pinned execution paths exercised in CI, not as turnkey reproduction of the case-study science.

Reviewer: I think this framing is a little confusing. The reproducibility contribution while it depends on inputs and the state, if the pipeline repo is also shared along with the package repo, it can backup the claim. reframing reproducibility as repeatable paths exercised in CI seems undersell

Assets without such idempotent output conventions recompute from the start on retry.
Recovering scheduler accounting metadata (`node-hours`, CPU efficiency) requires Slurm accounting (`sacct`) to be available on the site; where it is not, the job still runs and logs stream, but those metadata fields are omitted.

## The same code across targets

The integration is designed so that one asset definition runs in every target.
An asset calls `compute.run`, passes a payload path and its resource requirements, and yields the results.
The example below is the GPU extraction stage of the firm-discovery case study, abridged: imports and the partition and resource definitions are omitted, and `_payload`, `model_path`, and `country_crawl_partitions` are defined elsewhere in the project.
The asset states hardware requirements in deployment-neutral terms; each target maps those requirements to local execution, CI resources, or site-specific Slurm options.

```python
@dg.asset(partitions_def=country_crawl_partitions)
def cc_family_nuextract_results(
    context: dg.AssetExecutionContext,
    compute: ComputeResource,
):
    completed = compute.run(
        context=context,

>         payload_path=_payload("extract_nuextract.py"),

Reviewer: would it be possible to share the github repo here related to the pipeline run ?

Host names, transport, credentials, partition names, and site policy stay in deployment configuration, so the same asset source remains portable across targets.

For distributed workloads the Ray launcher [@moritz2018ray] replaces the Bash launcher in the same `ComputeResource`.
It allocates a multi-node Slurm job, starts a Ray head and workers across the allocation, runs the payload against that cluster, and terminates it.
The Bash and Ray launchers are the two production-stable launchers; a Spark launcher, session-based allocation reuse, and heterogeneous jobs are experimental and not covered by the operational experience reported here.

## Case study: web-scale firm discovery

We operate dagster-slurm on a pipeline that builds a structured corpus of European firms from public web text.
The scientific question is which firms present themselves as family-owned.
The engineering question, and the one this paper evaluates, is how to run a pipeline whose stages have very different hardware needs without maintaining two toolchains.

### What the pipeline computes

The pipeline runs once for each region and Common Crawl snapshot [@commoncrawl].

> The diagram below shows the stages.

Reviewer 2: ```suggestion
@fig:pipeline shows the stages.
```

> :::{figure} pipeline.png
> Stages of one (region, crawl) run.

Reviewer 2: ```suggestion
:::{figure} pipeline.png
:label: fig:pipeline
Stages of one (region, crawl) run.
```

Matching records are fetched from `s3://commoncrawl` with authenticated byte-range requests, cleaned with resiliparse [@resiliparse], and deduplicated to one page per domain and page type.
The extraction stage serves a fine-tuned NuExtract-2.0-8B model with vLLM [@nuextract; @kwon2023vllm] and returns a 30-field JSON record per page covering the company name, registration and VAT identifiers, legal form, address, described products, and an explicit family-ownership flag with a supporting sentence.
Records are consolidated to one row per firm and matched to a commercial company registry by identifier, domain, or name.

A single multilingual base model serves all target regions, reducing the need to train separate country-specific extractors.
Rather than train a separate extractor per country, we adapt the one 8.33-billion-parameter model per language group with a LoRA adapter [@hu2021lora] that trains 40.4 million parameters, 0.48% of the weights.

> The adapters are trained in a chain, each starting from the previous one, with a fraction of earlier-region data replayed; on held-out test domains the adapter is intended to retain coverage of earlier language groups while adding the new one, and our internal evaluation reports eval loss between 0.047 and 0.055 down the chain.

Reviewer 2: ```suggestion
The adapters are trained in a chain, each starting from the previous one, with a fraction of earlier-region data replayed.
On held-out test domains, extraction quality on earlier language groups changes by FILL IN (metric, before and after the next link) while the new group reaches FILL IN, indicating that the chain adds coverage without measurable regression.
```

If you do not have that measurement, delete the claim entirely rather than reporting the raw eval loss. A loss value with no task, scale, or baseline is not interpretable, and "is intended to retain coverage" is not a claim.

  - vLLM, gpu mem util 0.85; 15,000-22,000 pages/hour per A100
```

The pipeline is partitioned so that one (region, crawl) pair is one Slurm job.
The case-study results shown below cover fifteen regional groupings (DACH, plus the Nordic, Baltic, and Slavic countries); the pipeline's full partition space spans two axes, the registered country-code top-level domain (23 values) and the detected content language (19 values), each crossed with the configured set of 69 crawl snapshots, giving partition spaces of 1,587 and 1,311 (region, crawl) jobs respectively that are then merged and aggregated into the regional groupings.
Materializing a partition is independent and resumable: a relaunched partition reads the URLs already written and skips them, and a large region shards across GPUs by a hash of the URL so each worker takes a disjoint slice.

The filtering tier is a cost control, not only a convenience.
For the Romance languages, language does not equal country: Spanish and French reach across Latin America and Africa, so the index lists 1.25 billion Spanish and 1.11 billion French URLs, most of them out of scope.
These collapse to 6.9 million distinct domains.

> Even after deduplicating to one page per domain and page type, running extraction over the resulting candidate pages without first applying the registry country filter would exceed 1,100 A100-hours, with most of it spent on out-of-scope domains.

Reviewer 2: ```suggestion
Even after deduplicating to one page per domain and page type, roughly FILL IN million candidate pages remain; extracting over all of them at the 15,000 to 22,000 pages per hour of @tbl:tiers, without first applying the registry country filter, would exceed 1,100 A100-hours, with most of it spent on out-of-scope domains.
```

This is the strongest argument for the multi-tier design and currently the one number a reader cannot check.

:::

As an internal consistency check we measure eponymy, whether a firm's name contains a disclosed manager's or owner's surname, on the subset of firms that both name a person and have a company name.
Among that subset, 32% are eponymous overall; the rate is 52% for self-described family firms against 32% for the rest (@fig:eponymy).
This is a descriptive consistency check rather than ground truth: it can be confounded by region, sector, firm size, language, and extraction error, and we do not interpret it as validation of the family-ownership measure.

:::{figure} eponymy.png
:label: fig:eponymy
Eponymy rate, split by family-ownership self-description.
Sample restriction: firms that disclose at least one person and have a non-empty company name.
Numerator: firms whose company name contains a disclosed surname; denominator: firms in each group within that restricted sample.
The family vs non-family gap (52% vs 32%) is shown as a within-sample comparison, not a population estimate.
:::

These results are a by-product here.
The contribution this paper evaluates is the orchestration that produced them with one asset graph rather than two.

## Evaluation

> We assess the integration as a piece of research software along three dimensions: that it is tested, that it has been operated at production scale, and that it surfaces useful observability.

Reviewer 2: The evaluation covers tested, operated at scale, and observable, but not what the abstraction costs. 

* What is the median added latency per job from environment packaging, payload transfer, and Pipes polling, compared to a bare sbatch of the same payload? 
* And how much of that does the pre-deployed-environment path remove? 

You already have the instrumentation that produced 9,626 jobs of accounting data, and there are 2,500 words of headroom. 

Without this number the paper describes the system rather than evaluates it.

* - Package
  - dagster-slurm, Apache-2.0, github.com/ascii-supply-networks/dagster-slurm
```

**Reproducibility (testing).** The integration tests run in GitHub Actions against a containerized Slurm cluster.
CI provisions the environment with Pixi, deploys it once, starts a Slurm cluster in Docker, and runs Dagster assets [@docker] through the local and Slurm execution modes, exercising the same code path users run in production.
The CI fixture materializes the same test assets in local and Dockerized Slurm modes and compares their emitted Dagster metadata and output files.
The full build, unit, and Slurm-on-Docker integration job completes in about 16 minutes of wall time per commit, which keeps the cross-environment check cheap enough to run on every change.
This improves the repeatability of the selected execution paths, in line with current recommendations for CI-based HPC reproducibility [@hayotsasson2025address]; as noted above, it does not by itself reproduce a production scientific run.

**Operational experience.** The system has run the case-study pipeline in production on the TU Wien DataLAB cluster.

> Over the ten-week accounting window in @tbl:ops it submitted 9,626 Slurm jobs, of which 79.5% completed and 9.9% failed for reasons other than memory or time limits (with 64 out-of-memory and 57 timeout terminations counted separately; a further 9.3% were cancelled, which on this cluster includes deliberate reruns for stack consistency and retry supersessions rather than only errors), and it consumed 2,509 GPU-hours and 118,914 CPU core-hours.

Reviewer 2: ```suggestion
Over the ten-week accounting window in @tbl:ops it submitted 9,626 Slurm jobs and consumed 2,509 GPU-hours and 118,914 CPU core-hours.
Of those jobs, 79.5% completed, 9.9% failed for reasons other than memory or time limits, 9.3% were cancelled, 0.7% hit out-of-memory, and 0.6% hit the walltime limit.
The cancelled jobs are largely deliberate reruns for stack consistency and retry supersessions rather than errors.
The largest failure category resolves to FILL IN (top three causes, e.g. transient S3 fetch errors, model-server startup timeouts, node-local disk exhaustion), which the scheduler metadata surfaced in Dagster without additional instrumentation.
```

That last sentence turns the paper's weakest number into its best demonstration.

These totals are read directly from Slurm's accounting database (`sreport` and `sacct`), independent of Dagster's own run records; we use the scheduler as the source of truth precisely because the orchestrator's event log can drift when a connection is lost mid-job.
SSH ControlMaster fallbacks, password-based jump hosts, login-node hygiene, and queue, QoS, and reservation overrides are documented per site, with verification snippets using `squeue` and `scontrol`.
During development we also ran the integration against the VSC-5 and Leonardo supercomputers to validate portability; no production workloads were submitted to those sites, and their development-time accounting records are no longer available to us, so every measured figure in this paper is from DataLAB.

> We report this as production operational experience rather than as a scheduler-scalability benchmark, which would require a controlled study of submission throughput and concurrent-job limits that is outside this paper's scope.

Reviewer 2: ```suggestion
We report this as production operational experience rather than as a scheduler-scalability benchmark, which would require a controlled study of submission throughput and concurrent-job limits that is outside this paper's scope.

**Limitations.** Five constraints bound these results.
Data movement is left to the user, so cross-target portability of asset code depends on a deployment-mode-aware path convention rather than on anything the package enforces.
Scheduler accounting metadata requires `sacct` to be exposed at the site; where it is not, jobs run and logs stream but those fields are absent.
Clean restart depends on asset-level idempotency, and assets without an idempotent output convention recompute from the start.
Production measurements come from one cluster and one ten-week window, not from a multi-site or lifetime study.
The Spark launcher, session-based allocation reuse, and heterogeneous jobs are experimental and are not covered by any measurement reported here.
```

All five facts are already in the paper at lines 83, 101, 100, 329, and 166. 

Collecting them makes the paper's honesty a visible strength rather than something the reader has to assemble.

> **Observability.** Slurm job identifiers, CPU efficiency, memory use, and node-hours surface as Dagster metadata where Slurm accounting is available, and Ray clusters stream their output back through Pipes.
> Conventional Dagster asset checks and alerting therefore operate over those metadata and failure signals without modification, a property that separate script-based workflows often lose or must rebuild manually.

Reviewer 2: ```suggestion
**Observability.** Slurm job identifiers, CPU efficiency, memory use, and node-hours surface as Dagster metadata where Slurm accounting is available, and Ray clusters stream their output back through Pipes.
Across the accounting window, FILL IN% of submitted jobs returned a complete scheduler metadata record; the remainder ran and streamed logs but lacked `sacct` fields.
Conventional Dagster asset checks and alerting therefore operate over those metadata and failure signals without modification, a property that separate script-based workflows often lose or must rebuild manually.
```

Tested has CI timings and operated-at-scale has sreport totals. 

Observability is the one dimension of the three with nothing measured behind it.

Parsl, PSI/J, executorlib, Snakemake, and Nextflow provide mature scheduler-facing abstractions for functions, jobs, rules, or processes.
Flyte and Airflow provide broader workflow control planes, usually with a container-first deployment model.
dagster-slurm targets a narrower case: teams that already model their work as Dagster assets and want Slurm to be one execution target for those assets rather than a separate workflow boundary.

## Related work

The Python HPC workflow ecosystem is rich.
Parsl provides parallel scripting with multi-site execution and SSH submission [@parsl]; executorlib extends the standard `concurrent.futures` interface to schedulers with per-function resource control [@executorlib]; jobflow defines workflows with decorators [@jobflow]; and PSI/J offers a portable submission interface across schedulers [@psij].
In the bioinformatics tradition, Snakemake [@koster2012snakemake] and Nextflow [@ditommaso2017nextflow] define file-driven rule and process graphs that submit to Slurm through executors, and Dask [@rocklin2015dask] parallelizes array and dataframe computation, including onto Slurm via dask-jobqueue [@daskjobqueue].

> General-purpose orchestrators such as Apache Airflow [@airflow] and Flyte schedule task graphs but target containerized cloud backends first.

Reviewer: Would be better to add a Flyte reference similar to airflow
